client/client-multi/client-c/src/kaa/platform-impl/common/common_key_utils.c

```diff
@@ -0,0 +1,391 @@
+    mbedtls_pk_init(&local_ctx);
+    int err = set_rsa_public_key(&local_ctx, remote_key, remote_key_size);
+    if (err) {
+      goto exit;
+    }
+    err = rsa_encrypt(&local_ctx, keys.session_key,
+                      KAA_SESSION_KEY_LENGTH, keys.encrypted_session_key);
+    if (err) {
+      goto exit;
+    }
+
+    keys.encrypted_session_key_length = ENC_SESSION_KEY_LENGTH;
+
+    *buffer = keys.encrypted_session_key;
+    *buffer_size = keys.encrypted_session_key_length;
+
+exit:
+    mbedtls_pk_free(&local_ctx);
+    return err ? KAA_ERR_BADDATA : KAA_ERR_NONE;
+}
+
+__attribute__((weak))
+void ext_get_endpoint_public_key(uint8_t **buffer, size_t *buffer_size)
+{
+    if (!buffer || !buffer_size) {
+      return;
+    }
+
+    *buffer = kaa_rsa_public_key;
+    *buffer_size = KAA_RSA_PUBLIC_KEY_LENGTH;
+}
+
+void ext_get_encrypted_data_size(size_t input_size, size_t *output_size)
+{
+    if (!input_size || !output_size) {
+        *output_size = 0;
+        return;
+    }
+    *output_size = !(input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE) ?
+                   input_size + AES_ECB_ENCRYPTION_CHUNCK_SIZE :
+                   (AES_ECB_ENCRYPTION_CHUNCK_SIZE -
+                    (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE) + input_size);
+}
+
+kaa_error_t ext_encrypt_data(uint8_t *input, size_t payload_size, uint8_t *output)
+{
+    if (!input || !payload_size || !output) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    /* Adding PKCS7 padding */
+    size_t enc_data_size;
+    ext_get_encrypted_data_size(payload_size, &enc_data_size);
+    uint8_t padding = enc_data_size - payload_size;
+    memset(input + payload_size, padding, padding);
+
+    /* Process the buffer chunk by chunk */
+    while (enc_data_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_ENCRYPT, input,
+                                KAA_SESSION_KEY_LENGTH,
+                                output, keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+
+        enc_data_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t ext_decrypt_data(uint8_t *input, size_t input_size,
+                             uint8_t *output, size_t *output_payload_size)
+{
+    if (!input || !input_size || !output || !output_payload_size) {
+      return KAA_ERR_BADPARAM;
+    }
+
+    size_t tmp_in_size = input_size;
+    while (input_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_DECRYPT, input,
+                                KAA_SESSION_KEY_LENGTH, output,
+                                keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+        input_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    /* Reduce PKCS7 padding */
+    uint8_t padding_length = *(output - 1);
```

Comment:
Is it possible that there is no padding in buffer and last payload byte != 0? If yes, you will eventually interpret payload byte at index `output - 1` as padding length.